src/CustomElementInternals.js

```diff
@@ -0,0 +1,294 @@
+import CustomElementDefinition from './CustomElementDefinition';
+import * as Utilities from './Utilities';
+import * as CESymbols from './CustomElementInternalSymbols';
+import CEState from './CustomElementState';
+
+export default class CustomElementInternals {
+  constructor() {
+    /** @type {!Map<string, !CustomElementDefinition>} */
+    this._localNameToDefinition = new Map();
+
+    /** @type {!Map<!Function, !CustomElementDefinition>} */
+    this._constructorToDefinition = new Map();
+
+    /** @type {!Array<!function(!Node)>} */
+    this._patches = [];
+  }
+
+  /**
+   * @param {string} localName
+   * @param {!CustomElementDefinition} definition
+   */
+  setDefinition(localName, definition) {
+    this._localNameToDefinition.set(localName, definition);
+    this._constructorToDefinition.set(definition.constructor, definition);
+  }
+
+  /**
+   * @param {string} localName
+   * @return {!CustomElementDefinition|undefined}
+   */
+  localNameToDefinition(localName) {
+    return this._localNameToDefinition.get(localName);
+  }
+
+  /**
+   * @param {!Function} constructor
+   * @return {!CustomElementDefinition|undefined}
+   */
+  constructorToDefinition(constructor) {
+    return this._constructorToDefinition.get(constructor);
+  }
+
+  /**
+   * @param {!function(!Node)} listener
+   */
+  addPatch(listener) {
+    this._patches.push(listener);
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  patchTree(node) {
+    Utilities.walkDeepDescendants(node, node => this.patch(node));
```

Comment:
This can't be implemented as an additional treewalk of the entire world that typically (on good browsers with no "patches") does nothing.... too expensive.  Should be ok to piggy-back on `upgradeTree`.  Even then it seems a little too cute to have a pluggable patching facility-- do you have other use cases than `innerHTML` and `textContent`?  I'd prefer this be done as cheaply as possible and make sure it's viable before we start generalizing and abstracting.